File: code/sorting/insertion_sort/insertion_sort.swift
/* Part of Cosmos by OpenGenus Foundation */

//
//  insertion_sort.swift
//  Created by DaiPei on 2017/10/11.
//

import Foundation

func insertionSort(_ array: inout [Int]) {
    for i in 1..<array.count {
        var j = Int(i)
        let key = array[j]
        while j >= 1 && key < array[j - 1] {
            array[j] = array[j - 1]
            j -= 1
        }
        array[j] = key
    }
}

File: code/sorting/src/insertion_sort/insertion_sort.dart
main() {
  print(insertionSort([8, 9, 4, 2, 6, 10, 12]));
}

List<int> insertionSort(List<int> list) {
  for (int j = 1; j < list.length; j++) {
    int key = list[j];
    int i = j - 1;
    while (i >= 0 && list[i] > key) {
      list[i + 1] = list[i];
      i = i - 1;
      list[i + 1] = key;
    }
  }
  return list;
}
